### PR TITLE
[MOON-1460] change eligibility ratio to be a number

### PR DESCRIPTION
**Pull request by nbaztec**
Updates eligibility ratio to be a number to handle cases of more than `100` authors

### CODE REVIEW

File: pallets/author-slot-filter/src/lib.rs [hunk 1]
```diff
@@ -128,54 +132,60 @@ pub mod pallet {
...
 			<Pallet<T>>::deposit_event(Event::EligibleUpdated(new));
 
 			Ok(Default::default())
 		}
 	}
 
+	/// The type of eligibility to use
+	pub type EligibilityValue = u32;
+
 	/// The percentage of active authors that will be eligible at each height.
```
Comment (librelois) [inline pallets/author-slot-filter/src/lib.rs]: This comment should be updated too

File: pallets/author-slot-filter/src/lib.rs [hunk 2]
```diff
@@ -128,54 +132,60 @@ pub mod pallet {
...
+	/// Total number of eligible authors
+	const TOTAL_ELIGIBLE_AUTHORS: EligibilityValue = 100;
 
-	// Default value for the `EligibleRatio` is one half.
+	// Default value for the `EligibleCount`.
 	#[pallet::type_value]
-	pub fn Half<T: Config>() -> Percent {
-		Percent::from_percent(50)
+	pub fn Half<T: Config>() -> EligibilityValue {
+		div_ceil(TOTAL_ELIGIBLE_AUTHORS, 2)
```
Comment (librelois) [inline pallets/author-slot-filter/src/lib.rs]: Create a new constant HALF_TOTAL_ELIGIBLE_AUTHORS instead.
It's make no sense to compute a constant value at runtime.

File: pallets/author-slot-filter/src/lib.rs [hunk 3]
```diff
@@ -128,54 +132,60 @@ pub mod pallet {
...
-		pub eligible_ratio: Percent,
+		pub eligible_count: EligibilityValue,
 	}
 
 	#[cfg(feature = "std")]
 	impl Default for GenesisConfig {
 		fn default() -> Self {
 			Self {
-				eligible_ratio: Percent::from_percent(50),
+				eligible_count: div_ceil(TOTAL_ELIGIBLE_AUTHORS, 2),
```
Comment (librelois) [inline pallets/author-slot-filter/src/lib.rs]: Create a new constant HALF_TOTAL_ELIGIBLE_AUTHORS instead.
It's make no sense to compute a constant value at runtime.

**Review by librelois — CHANGES_REQUESTED** [review]

File: pallets/author-slot-filter/src/lib.rs [hunk 4]
```diff
@@ -128,54 +131,63 @@ pub mod pallet {
...
-	/// The percentage of active authors that will be eligible at each height.
+	/// The type of eligibility to use
+	pub type EligibilityValue = u32;
+
+	/// The number of active authors that will be eligible at each height.
 	#[pallet::storage]
-	#[pallet::getter(fn eligible_ratio)]
-	pub type EligibleRatio<T: Config> = StorageValue<_, Percent, ValueQuery, Half<T>>;
+	#[pallet::getter(fn eligible_count)]
+	pub type EligibleCount<T: Config> = StorageValue<_, EligibilityValue, ValueQuery, Half<T>>;
```
Comment (librelois) [inline pallets/author-slot-filter/src/lib.rs]: This implies that a migration is necessary, it must also be decided how the percentage is translated into number during the migration, @alan do you have an idea on this?

**Review by librelois — CHANGES_REQUESTED** [review]

File: pallets/author-slot-filter/src/lib.rs [hunk 5]
```diff
@@ -128,54 +131,63 @@ pub mod pallet {
...
+			EligibleCount::<T>::put(&new);
 			<Pallet<T>>::deposit_event(Event::EligibleUpdated(new));
 
 			Ok(Default::default())
 		}
 	}
 
-	/// The percentage of active authors that will be eligible at each height.
+	/// The type of eligibility to use
+	pub type EligibilityValue = u32;
```
Comment (librelois) [inline pallets/author-slot-filter/src/lib.rs]: Use `NonZeroU32` instead to ensure at compile time that it can't be 0

File: pallets/author-slot-filter/src/lib.rs (hunk 4 above)
Comment (girazoki) [reply pallets/author-slot-filter/src/lib.rs]: Since this might affect block production I suggest we keep the previous storage, and add the new EligibleCount in a new storage. We can in the future remove _EligibleRatio_

Comment (crystalin) [reply pallets/author-slot-filter/src/lib.rs]: Why do we want to keep the previous storage? How is it helping ?

**Comment by girazoki** [comment]
I am having trouble figuring out why this brings and advantage over what was before. I dont understand why the previous approach could not handle more than 100 authors, would you mind explaining?

**Comment by nbaztec** [comment]
I think it was mentioned that if the total author count was `> 100` say `150` a value of `50% = 75` would be erroneous since our max slots are always 100 (at least that's my understanding)

File: pallets/author-slot-filter/src/lib.rs [hunk 6]
```diff
@@ -34,9 +34,54 @@ mod benchmarks;
...
+pub struct NonZeroU32(u32);
+
+impl core::ops::Deref for NonZeroU32 {
+	type Target = u32;
+	fn deref(&self) -> &Self::Target {
+		&self.0
+	}
+}
+
+impl parity_scale_codec::EncodeLike<u32> for NonZeroU32 {}
```
Comment (librelois) [inline pallets/author-slot-filter/src/lib.rs]: Since an instance of your type `NonZeroU32` can be created by SCALE decoding, this type can be 0, so we don't have the desired guarantee. 
Instead, `Decode` should be implemented by hand to return an error if the storage contains zero.

File: pallets/author-slot-filter/src/lib.rs (hunk 4 above)
Comment (girazoki) [reply pallets/author-slot-filter/src/lib.rs]: In general I am hesitant of migrating a storage prefix involved in block production without keeping the previous storage. I think that has been our approach so far with every data migration we have performed if we suspected it could affect block production

**Comment by girazoki** [comment]
Ok got it thanks to @librelois, it is a precision issue right?

File: pallets/author-slot-filter/src/lib.rs (hunk 4 above)
Comment (crystalin) [reply pallets/author-slot-filter/src/lib.rs]: Ok, we could mark it as deprecated. But I don't think it makes anything safer in terms of block production

File: pallets/author-slot-filter/src/migration.rs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+			migration::put_storage_value::<Option<NonZeroU32>>(
+				PALLET_NAME,
+				NEW_ITEM_NAME,
+				&[],
+				NonZeroU32::new(new_value),
+			);
+		}
+
+		let db_weights = T::DbWeight::get();
+		db_weights.write + db_weights.read
```
Comment (girazoki) [inline pallets/author-slot-filter/src/migration.rs]: Since you are writing only inside the if condition, you can only add the db_weights.write thing if you enter the if condition

File: pallets/author-slot-filter/src/migration.rs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+use frame_support::traits::OnRuntimeUpgrade;
+use frame_support::weights::Weight;
+use sp_runtime::Percent;
+
+use super::num::NonZeroU32;
+use super::pallet::Config;
+
+pub struct EligibleRatioToEligiblityCount<T>(PhantomData<T>);
+
+const PALLET_NAME: &[u8] = b"AuthorSlotFilter";
```
Comment (girazoki) [inline pallets/author-slot-filter/src/migration.rs]: I would put these inside the migration. Or at least change the naming. In the future OLD_ITEM_NAME might refer to multiple things

File: pallets/author-slot-filter/src/migration.rs [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+			);
+		}
+
+		let db_weights = T::DbWeight::get();
+		db_weights.write + db_weights.read
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<(), &'static str> {
+		Ok(())
```
Comment (girazoki) [inline pallets/author-slot-filter/src/migration.rs]: Here you should check if something exists in the previous storage value, and if so, write the value in a temp value that will be later read in the post_upgrade state

File: pallets/author-slot-filter/src/migration.rs [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn post_upgrade() -> Result<(), &'static str> {
+		Ok(())
```
Comment (girazoki) [inline pallets/author-slot-filter/src/migration.rs]: Here, if the temp variable you wrote before exists, you should verify that the new storage value is what you expect based on the value in the temp variable

File: pallets/author-slot-filter/src/migration.rs (hunk 8 above)
Comment (nbaztec) [reply pallets/author-slot-filter/src/migration.rs]: aren't these already in the migrations module? \
Alternatively maybe we could use `ELIGIBLE_RATIO_ITEM_NAME` & `ELIGIBLE_COUNT_ITEM_NAME`?

File: pallets/author-slot-filter/src/migration.rs (hunk 9 above)
Comment (nbaztec) [reply pallets/author-slot-filter/src/migration.rs]: By previous values, do you mean of both `EligibilityRatio` and `EligibilityCount`? So we'd check that the old `Percent` value existed for `EligibilityRatio` and empty for `EligibilityCount`?

Comment (girazoki) [reply pallets/author-slot-filter/src/migration.rs]: No, just EligibilityRatio. If the storage is non-empty, you write its value into a temp value

File: pallets/author-slot-filter/src/migration.rs (hunk 8 above)
Comment (girazoki) [reply pallets/author-slot-filter/src/migration.rs]: yeah, that works better!

File: pallets/author-inherent/src/lib.rs [hunk 11]
```diff
@@ -218,154 +218,3 @@ impl InherentError {
 		}
 	}
 }
-
-#[cfg(test)]
-mod tests {
```
Comment (girazoki) [inline pallets/author-inherent/src/lib.rs]: Why did these dissapear?

File: pallets/author-slot-filter/src/lib.rs [hunk 12]
```diff
@@ -259,10 +288,13 @@ pub mod tests {
...
+			let value = num::NonZeroU32::new(34);
 
-			assert_ok!(AuthorSlotFilter::set_eligible(Origin::root(), percent));
-			assert_eq!(AuthorSlotFilter::eligible_ratio(), percent)
+			assert_ok!(AuthorSlotFilter::set_eligible(
+				Origin::root(),
+				value.clone()
+			));
+			assert_eq!(AuthorSlotFilter::eligible_count(), value)
 		});
```
Comment (girazoki) [inline pallets/author-slot-filter/src/lib.rs]: Can you add one more test which tests the migration? You can take https://github.com/PureStake/moonbeam/blob/c582e2f34a2271a22e60a73dbc6d4ffde90de837/pallets/xcm-transactor/src/tests.rs#L319 as a reference

File: pallets/author-inherent/src/lib.rs (hunk 11 above)
Comment (nbaztec) [reply pallets/author-inherent/src/lib.rs]: Removed in https://github.com/PureStake/nimbus/pull/37 after discussion with team.

tl;dr they were broken/uncompilable and the system under test didn't even exist in code.

Comment (girazoki) [reply pallets/author-inherent/src/lib.rs]: I see, sorry wasnt aware!

File: pallets/author-slot-filter/src/lib.rs (hunk 12 above)
Comment (nbaztec) [reply pallets/author-slot-filter/src/lib.rs]: added, please review.

**Review by girazoki — APPROVED** [review]
LGTM

File: pallets/author-slot-filter/src/lib.rs [hunk 13]
```diff
@@ -128,20 +142,26 @@ pub mod pallet {
...
+			EligibleCount::<T>::put(&new);
 			<Pallet<T>>::deposit_event(Event::EligibleUpdated(new));
 
 			Ok(Default::default())
 		}
 	}
 
-	/// The percentage of active authors that will be eligible at each height.
+	/// The type of eligibility to use
+	pub type EligibilityValue = Option<NonZeroU32>;
```
Comment (librelois) [inline pallets/author-slot-filter/src/lib.rs]: why Option ?

Comment (nbaztec) [reply pallets/author-slot-filter/src/lib.rs]: I kept the same interface as provided by https://docs.substrate.io/rustdocs/latest/sp_std/num/struct.NonZeroU32.html#method.new to make it easier to port once I file a couple of upstream PRs to substrate. `new` checks if the value is `> 0` putting it in `Some(...)`, else returning a `None`.

Comment (librelois) [reply pallets/author-slot-filter/src/lib.rs]: I not talking about the new() signature, but about the storage type, the storage should not contains an Option, it's useless because all storage values are already optional, and you already provided a default value in case the storage item is empty.

Comment (librelois) [reply pallets/author-slot-filter/src/lib.rs]: So please remove this Option from the type used for storage.

File: pallets/author-slot-filter/src/num.rs [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+		Self(n)
+	}
+
+	/// Returns the the underlying number
+	pub fn get(&self) -> u32 {
+		self.0
+	}
+}
+
+impl Serialize for NonZeroU32 {
```
Comment (librelois) [inline pallets/author-slot-filter/src/num.rs:62]: Serialize and Deserialize impls should be compiled with std feature only

**Review by librelois — CHANGES_REQUESTED** [review]
I just have 2 blocking comments, overall it's good :)

File: pallets/author-slot-filter/src/lib.rs [hunk 15]
```diff
@@ -150,119 +168,46 @@ pub mod pallet {
 		Percent::from_percent(50)
 	}
 
+	/// The number of active authors that will be eligible at each height.
+	#[pallet::storage]
+	#[pallet::getter(fn eligible_count)]
+	pub type EligibleCount<T: Config> =
+		StorageValue<_, EligibilityValue, ValueQuery, DefaultEligibilityValue<T>>;
+
+	/// Default total number of eligible authors, not NOT be 0.
```
Comment (librelois) [inline pallets/author-slot-filter/src/lib.rs]:
```suggestion
	/// Default total number of eligible authors, must NOT be 0.
```

File: pallets/author-slot-filter/src/migration.rs [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+		if let Some(old_value) =
+			migration::get_storage_value::<Percent>(PALLET_NAME, ELIGIBLE_RATIO_ITEM_NAME, &[])
+		{
+			let total_authors = <T as Config>::PotentialAuthors::get().len();
+			let new_value: u32 = percent_of_num(old_value, total_authors as u32);
+			migration::put_storage_value(
+				PALLET_NAME,
+				ELIGIBLE_COUNT_ITEM_NAME,
+				&[],
+				NonZeroU32::new_unchecked(new_value),
```
Comment (librelois) [inline pallets/author-slot-filter/src/migration.rs]: If you use `new_unchecked` you need to ensure that new_value > 0, and apply a minimum (`1`) if needed

**Review by librelois — APPROVED** [review]
